Track user reviews

bnmajor opened the pull request:
This PR adds support for tracking reviews. Currently the labels annotation metadata is updated to track which superpixels have been reviewed, what their new value is, who did that review, as well as the date and time of the review. On retraining the labels are updated to the reviewed values (if there has been a change).

Relies on changes in #130

bnmajor commented:
Moved to this branch for follow-up ([See original comment](https://github.com/DigitalSlideArchive/wsi-superpixel-guided-labeling/pull/101#issuecomment-2256221102)):

> Once one selects a class in the change class dropdown, would this happen immediately?

Yes

> In which case if you are filtering by class would these chips vanish? Or would we have a confirm button?

Reviewed chips are not automatically hidden, so nothing would vanish unless they have filtered to only show chips with no review. In that case they would vanish immediately. I was kind of considering the "retrain" (or maybe a "complete reviews" button) to be confirmation at the end but if a "confirm" button would be useful we could discuss that as well!

bnmajor commented:
Some additional notes that may warrant further discussion:
- All previous labels are made available for review. There is currently no way of knowing if a previous label was reviewed and then re-trained on but we may need to (or eventually want to) account for this.
- Labels for review are compared to current predictions. If there is a need to collect all labels and predictions from all previous epochs and compare labeled to predicted on a per-epoch level that would be something we need to discuss further.
- Reviews do not affect the guided mode selections and are not actually applied until a re-training is run.
- Review selections (filters, sort, view, etc) are not remembered between sessions. If there is a need we can discuss which selections might make sense to save to the config file to be remembered between sessions.
- Group/filter/sort etc. is done on the user selection, not the reviewer's selection.

bnmajor commented:
Updating with answers from our last meeting:

> Some additional notes that may warrant further discussion:
> 
> * All previous labels are made available for review. There is currently no way of knowing if a previous label was reviewed and then re-trained on but we may need to (or eventually want to) account for this.

We do not want to review an already-reviewed, unchanged superpixel. By default now, reviewed superpixels are filtered out. The filters menu now also contain options to see superpixels that the current user has reviewed, those reviewed by others, and those that have not been reviewed at all.

> * Reviews do not affect the guided mode selections and are not actually applied until a re-training is run.

This is acceptable for now.

> * Group/filter/sort etc. is done on the user selection, not the reviewer's selection.

According to previous discussion this should be the correct behavior

bnmajor commented:
One piece that I still need some clarification on relates to the following:

> * Labels for review are compared to current predictions. If there is a need to collect all labels and predictions from all previous epochs and compare labeled to predicted on a per-epoch level that would be something we need to discuss further.

As an example - given two users, Junior and Senior:

1. Junior labels a `superpixel N` as `category A`.
2. Senior reviews all labels and changes `superpixel N` from `category A` to `category B` and re-runs training.
3. The current epoch now says that the user selection for `superpixel N` is `category B` because that is what it was updated to before training so that it would train on the correct label.

Is this the expected behavior? Or should the user selection for `superpixel N` be reported as `category A`, like Junior originally labeled?

bnmajor commented:
> One piece that I still need some clarification on relates to the following:
> 
> > * Labels for review are compared to current predictions. If there is a need to collect all labels and predictions from all previous epochs and compare labeled to predicted on a per-epoch level that would be something we need to discuss further.
> 
> As an example - given two users, Junior and Senior:
> 
> 1. Junior labels a `superpixel N` as `category A`.
> 2. Senior reviews all labels and changes `superpixel N` from `category A` to `category B` and re-runs training.
> 3. The current epoch now says that the user selection for `superpixel N` is `category B` because that is what it was updated to before training so that it would train on the correct label.
> 
> Is this the expected behavior? Or should the user selection for `superpixel N` be reported as `category A`, like Junior originally labeled?

For reference:
Q: (TL;DR) Is it important to differentiate between a "user" and a "reviewer" selection once the selection has been trained on?
A: No

manthey commented:
Hmm... I'm on master with this branch merged in.  I started a new two image project, created three categories and labeled samples, trained, and used the hotkeys to set 9 chip values.  I then click Review and get a mostly empty area.  Is the automatic merge with master bad in some way?

![image](https://github.com/user-attachments/assets/42e2cae5-6bd9-4a79-a157-baca222a5319)

bnmajor commented:
> Hmm... I'm on master with this branch merged in. I started a new two image project, created three categories and labeled samples, trained, and used the hotkeys to set 9 chip values. I then click Review and get a mostly empty area. Is the automatic merge with master bad in some way?
> 
> ![image](https://private-user-images.githubusercontent.com/8781639/361531006-42e2cae5-6bd9-4a79-a157-baca222a5319.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2OTcxNjIsIm5iZiI6MTcyNDY5Njg2MiwicGF0aCI6Ii84NzgxNjM5LzM2MTUzMTAwNi00MmUyY2FlNS02YmQ5LTRhNzktYTE1Ny1iYWNhMjIyYTUzMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjZUMTgyNzQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODZlY2Y1MGI5ZGVhNzhhOWUzMjIyMTMxMjI4YTExOTMyZTVlMGFlNjBiYTk1MjQ5NjgzNWM3ZDI4OTJjOTQxMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Hu2HdkWmquvQhO4mf3Mo2K55y9f4_CP2VPt_OYDWYgc)

Given that there are no merge conflicts I'm more inclined to believe that there is a bug in the filtering instead... Can you try clearing all the filters to see if that fixes it?

manthey commented:
> Given that there are no merge conflicts I'm more inclined to believe that there is a bug in the filtering instead... Can you try clearing all the filters to see if that fixes it?

When I clear the filters (or filter by anything other than review), I see exactly one superpixel.  I also see a console error:
```
plugin.min.js?_=1724692876901:7 TypeError: Failed to execute 'unobserve' on 'IntersectionObserver': parameter 1 is not of type 'Element'.
    at o.filteredSortedGroupedSuperpixels (plugin.min.js?_=1724692876901:7:113843)
    at Ve (plugin.min.js?_=1724692876901:7:11219)
    at hn.run (plugin.min.js?_=1724692876901:7:27618)
    at fn (plugin.min.js?_=1724692876901:7:25590)
    at Array.<anonymous> (plugin.min.js?_=1724692876901:7:12257)
    at Ke (plugin.min.js?_=1724692876901:7:11702)
```

bnmajor commented:
> > Given that there are no merge conflicts I'm more inclined to believe that there is a bug in the filtering instead... Can you try clearing all the filters to see if that fixes it?
> 
> When I clear the filters (or filter by anything other than review), I see exactly one superpixel. I also see a console error:
> 
> ```
> plugin.min.js?_=1724692876901:7 TypeError: Failed to execute 'unobserve' on 'IntersectionObserver': parameter 1 is not of type 'Element'.
>     at o.filteredSortedGroupedSuperpixels (plugin.min.js?_=1724692876901:7:113843)
>     at Ve (plugin.min.js?_=1724692876901:7:11219)
>     at hn.run (plugin.min.js?_=1724692876901:7:27618)
>     at fn (plugin.min.js?_=1724692876901:7:25590)
>     at Array.<anonymous> (plugin.min.js?_=1724692876901:7:12257)
>     at Ke (plugin.min.js?_=1724692876901:7:11702)
> ```

The observer it is then, thanks! I'll test and ping you when it is fixed.

bnmajor commented:
@manthey This should be fixed now.

manthey commented:
If I don't include a review filter, the display looks correct, but applying any review filter (No review, by me, by others) I don't see any chips and I still see a console error:
```
plugin.min.js?_=1724852231236:7 TypeError: Failed to execute 'unobserve' on 'IntersectionObserver': parameter 1 is not of type 'Element'.
    at o.filteredSortedGroupedSuperpixels (plugin.min.js?_=1724852231236:7:113843)
    at Ve (plugin.min.js?_=1724852231236:7:11219)
    at hn.run (plugin.min.js?_=1724852231236:7:27618)
    at fn (plugin.min.js?_=1724852231236:7:25590)
    at Array.<anonymous> (plugin.min.js?_=1724852231236:7:12257)
    at Ke (plugin.min.js?_=1724852231236:7:11702)
```

manthey commented:
I have managed to confuse myself or there is an issue.  I took my previous example and added a fourth category "Artifacts".  In the review panel, I selected "no review" and "background", then Bulk Review.  I checked 7 or so chips then selected "Change Class" and changed to "Artifact" (I had not previously labelled any artifacts class).  Now, selecting "Reviewed by me" as the only filter, I see those chips still reporting "Background" (and, oddly, some chips that don't have the review marker):

![image](https://github.com/user-attachments/assets/f57ea5a6-152f-471c-be2e-250a75641abe)

bnmajor commented:
> I have managed to confuse myself or there is an issue. I took my previous example and added a fourth category "Artifacts". In the review panel, I selected "no review" and "background", then Bulk Review. I checked 7 or so chips then selected "Change Class" and changed to "Artifact" (I had not previously labelled any artifacts class). Now, selecting "Reviewed by me" as the only filter, I see those chips still reporting "Background" (and, oddly, some chips that don't have the review marker)

I was able to reproduce similar weirdness which boils down to the reviewer information not being updated everywhere. After the latest commit, copying your exact steps now I see the expected results.

manthey commented:
I still puzzled by what happens when I change a class during a review.  I can see that I've changed the class (though the superpixels still filter to the original class), but the class counts don't reflect the change and retraining acts like the superpixels are still the original class.

I was expecting the re-classified values to act as if I manually labelled them as the new class value.  For instance
![image](https://github.com/user-attachments/assets/e1cc6cb6-bec0-4733-b67c-971e7274dc72)
I would expect the "Artifact" entries would no longer be "Background".

manthey commented:
On a new project, when I click "Begin Training", I see a request to put the annotation that throws an error because the labels include None values rather than just integers.  I was able to do this twice.

manthey commented:
I also see a `applyReview is not defined` console warning.

bnmajor commented:
Additional changes since requests in last week's meeting:

- We were already tracking metadata for each review applied (user, date & time, category). We now also track all of this for labeled superpixels as well.
- We now track the epoch associated with each label and each review in the metadata. This is useful for some of the filters that we need, as well as keeping the label count badge accurate for the current epoch only (rather than a cumulative result).

Some questions I still have:

1. Is it useful to support further filtering on hiding already reviewed superpixels? Specifically I am wondering if it would be useful to be able to select between filtering out those that have been reviewed and trained on vs. those that have been reviewed this epoch but not yet trained on (or filtering out all reviewed, regardless of trained status). Personally I do not like that chips disappear as I review them but this may be a non-issue for actual users
3. Should we re-think/get some UI design input for filtering options? The number of items to filter on continues to grow, and there is a need to filter on user which could be a very long list. For now maybe we just have two filter menus - one for current filters, one for users?
    1. Related: at the moment we return the intersection of all values returned when multiple filters are applied. Is this correct, or should we return the union?

bnmajor commented:
> On a new project, when I click "Begin Training", I see a request to put the annotation that throws an error because the labels include None values rather than just integers. I was able to do this twice.

Trying to apply reviews that do not exist is no good... Fixed now.

> I also see a `applyReview is not defined` console warning.

Not sure what I did with all my import statements... Fixed now. Also renamed now to reflect the more general use case of the function (`applyReview`->`updateMetadata`)

manthey commented:
These issues may have nothing to do with the current branch and can be put off for a different PR:
- Sometimes when showing / hiding predictions the button seems to do nothing.  This happens erratically; I haven't found a consistent way to reproduce.
- We have some console errors `Cannot read properties of null (reading 'layers')` which probably occurs when switching images.  We should probably just guard this so it doesn't happen. 
- Possible coupled with the previous, sometimes I see a `Cannot read properties of null (reading 'removeEventListener')` error when we do `document.querySelector(".resize-handle").removeEventListener("mousedown", ...`

manthey commented:
Also, when labelling chips, I would expect the category counts to update -- they update when switching to review mode and back.

manthey reviewed (approved):
This appears to all be working well; any issues can be addressed in future PRs.
